pointmatcher/DataPointsFilters/Geometry.cpp

```diff
@@ -0,0 +1,170 @@
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL ETH-ASL BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+#include "Geometry.h"
+
+// Eigenvalues
+#include "Eigen/QR"
+#include "Eigen/Eigenvalues"
+
+#include "PointMatcherPrivate.h"
+#include "IO.h"
+#include "MatchersImpl.h"
+
+#include <boost/format.hpp>
+
+
+#include "utils.h"
+
+// SurfaceNormalDataPointsFilter
+// Constructor
+template<typename T>
+GeometryDataPointsFilter<T>::GeometryDataPointsFilter(const Parameters& params):
+	PointMatcher<T>::DataPointsFilter("GeometryDataPointsFilter",
+	        GeometryDataPointsFilter::availableParameters(), params),
+    keepUnstructureness(Parametrizable::get<int>("keepUnstructureness")),
+    keepStructureness(Parametrizable::get<T>("keepStructureness"))
+{
+}
+
+// Compute
+template<typename T>
+typename PointMatcher<T>::DataPoints
+GeometryDataPointsFilter<T>::filter(
+	const DataPoints& input)
+{
+	DataPoints output(input);
+	inPlaceFilter(output);
+	return output;
+}
+
+// In-place filter
+template<typename T>
+void GeometryDataPointsFilter<T>::inPlaceFilter(
+	DataPoints& cloud)
+{
+	typedef typename DataPoints::View View;
+	typedef typename DataPoints::Label Label;
+	typedef typename DataPoints::Labels Labels;
+
+	const int pointsCount(cloud.features.cols());
+	const int descDim(cloud.descriptors.rows());
+	const unsigned int labelDim(cloud.descriptorLabels.size());
+
+	// Check that the required eigenValue descriptor exists in the pointcloud
+    if (!cloud.descriptorExists("eigValues"))
+    {
+        throw InvalidField("GeometryDataPointsFilter: Error, no eigValues found in descriptors.");
+    }
+
+    // Validate descriptors and labels
+	int insertDim(0);
+	for(unsigned int i = 0; i < labelDim ; ++i)
+		insertDim += cloud.descriptorLabels[i].span;
+	if (insertDim != descDim)
+		throw InvalidField("SurfaceNormalDataPointsFilter: Error, descriptor labels do not match descriptor data");
+
+	// Reserve memory for new descriptors
+	const int dimSphericality(1);
+	const int dimUnstructureness(1);
+	const int dimStructureness(1);
+
+	boost::optional<View> sphericality;
+	boost::optional<View> unstructureness;
+	boost::optional<View> structureness;
```

Comment:
What is the advantage of these optionals? How are they mapped in memory?

Considering that libpointmatcher uses very flat representations (Eigen matrices), and speed is important, I would consider using a fixed size array. I assume that would make the resource usage of this filter more predictable.

Reply:
I've taken this scheme from another filter from the libpointmatcher (sampling surface normal). I understand the "optional" construct as a slightly smarter pointer. It points to a view on the data in the pointcloud, whose structure I don't want to change. Given that in the end all of these are pointers, I don't think there is any significant overhead (the expensive thing is done once in the `cloud.allocateDescriptors(cloudLabels);` call).